flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -99,91 +75,379 @@ class DoConcurrentConversion : public mlir::OpConversionPattern<fir::DoLoopOp> {
+        mlir::omp::VariableCaptureKind::ByRef;
+
+    if (fir::isa_trivial(eleType) || fir::isa_char(eleType)) {
+      captureKind = mlir::omp::VariableCaptureKind::ByCopy;
+    } else if (!fir::isa_builtin_cptr_type(eleType)) {
+      mapFlag |= llvm::omp::OpenMPOffloadMappingFlags::OMP_MAP_TO;
+      mapFlag |= llvm::omp::OpenMPOffloadMappingFlags::OMP_MAP_FROM;
     }
-    // ==== TODO (2) End ====
-    //
-    // TODO (1 & 2): Isolating analyses proposed in both TODOs, I think we can
-    // more easily generalize the pass to work for targets other than OpenMP,
-    // e.g. OpenACC, I think can, can reuse the results of the analyses and only
-    // change the code-gen/rewriting.
 
-    mlir::IRMapping mapper;
+    llvm::SmallVector<mlir::Value> boundsOps;
+    genBoundsOps(rewriter, liveIn.getLoc(), declareOp, boundsOps);
+
+    return Fortran::lower::omp::createMapInfoOp(
+        rewriter, liveIn.getLoc(), declareOp.getBase(), /*varPtrPtr=*/{},
+        declareOp.getUniqName().str(), boundsOps, /*members=*/{},
+        /*membersIndex=*/mlir::DenseIntElementsAttr{},
+        static_cast<
+            std::underlying_type_t<llvm::omp::OpenMPOffloadMappingFlags>>(
+            mapFlag),
+        captureKind, liveInType);
+  }
+
+  mlir::omp::TargetOp genTargetOp(mlir::Location loc,
+                                  mlir::ConversionPatternRewriter &rewriter,
+                                  mlir::IRMapping &mapper,
+                                  llvm::ArrayRef<mlir::Value> liveIns,
+                                  mlir::omp::TargetClauseOps &clauseOps) const {
+    auto targetOp = rewriter.create<mlir::omp::TargetOp>(loc, clauseOps);
+
+    genBodyOfTargetOp(rewriter, targetOp, liveIns, clauseOps.mapVars, mapper);
+    return targetOp;
+  }
+
+  void genBodyOfTargetOp(mlir::ConversionPatternRewriter &rewriter,
+                         mlir::omp::TargetOp targetOp,
+                         llvm::ArrayRef<mlir::Value> liveIns,
+                         llvm::ArrayRef<mlir::Value> liveInMapInfoOps,
+                         mlir::IRMapping &mapper) const {
+    mlir::Region &region = targetOp.getRegion();
+
+    llvm::SmallVector<mlir::Type> liveInTypes;
+    llvm::SmallVector<mlir::Location> liveInLocs;
+
+    for (mlir::Value liveIn : liveIns) {
+      liveInTypes.push_back(liveIn.getType());
+      liveInLocs.push_back(liveIn.getLoc());
+    }
+
+    rewriter.createBlock(&region, {}, liveInTypes, liveInLocs);
+
+    for (auto [arg, mapInfoOp] :
+         llvm::zip_equal(region.getArguments(), liveInMapInfoOps)) {
+      auto miOp = mlir::cast<mlir::omp::MapInfoOp>(mapInfoOp.getDefiningOp());
+      hlfir::DeclareOp liveInDeclare = genLiveInDeclare(rewriter, arg, miOp);
+      mapper.map(miOp.getVariableOperand(0), liveInDeclare.getBase());
+    }
+
+    auto terminator =
+        rewriter.create<mlir::omp::TerminatorOp>(targetOp.getLoc());
+    rewriter.setInsertionPoint(terminator);
```

Comment:
Nit: It's a commonly used pattern, so I think we don't need to split it into two statements.
```suggestion
    rewriter.setInsertionPoint(rewriter.create<mlir::omp::TerminatorOp>(targetOp.getLoc()));
```